### PR TITLE
LLM finetune sparsify masking

### PR DESCRIPTION
**Pull request by dsikka**
## Summary:

Add helper functions to the `finetune` task which allows us to finetune previously sparsified models. This is done by attaching masks using buffers after the sparsified model is loaded, to any weight which has already been pruned. A Composer callback function is also added to apply the masks to the weights, after each step during training. These helper functions will be converted to be applied through a recipe in a follow-up PR.

## Package Updates

llm-foundry has a requirement for `accelerate` which is incompatible with our transformers. Therefore, this moves `sparseml[transformers]` to the `nm` tag instead of making it a required installation. This allows us to install the correct `accelerate` for `llm-foundry` with the `llm` tag and the correct `accelerate` for `sparseml[transformers]` with the `nm` tag. Also, because our requirements are broad, there is an issue with older nightly versions. `setup.py` has been updated to look at more recent versions.

## Testing:

Testing using a 70% sparsified mpt model. The following command and config was used for this model:

```
max_seq_len: 2048
global_seed: 17
model_name_or_path: mosaicml/mpt-7b-instruct
model_name: ${model_name_or_path}
load_path:  /home/dsikka/mpt_7b_instruct_oneshot_sp70.pt
precision: amp_bf16

max_duration: 1ep # run for 2 epochs
eval_interval: 1ep
# eval_subset_num_batches: 3  # use this for quick testing
eval_first: true
seed: ${global_seed}

device_train_microbatch_size: 1 # set to catch potential OOM cuda errors
device_train_batch_size: 4
device_eval_batch_size: 4

# Run Name
run_name: testing_run

model:
  name: hf_causal_lm
  pretrained: true
  pretrained_model_name_or_path: ${model_name}
  max_seq_len: ${max_seq_len}

# Tokenizer
tokenizer:
  name: ${model_name}
  kwargs:
    model_max_length: ${max_seq_len}

# Dataloaders
train_loader:
  name: finetuning
  dataset:
    hf_name: mosaicml/dolly_hhrlhf
    split: train
    max_seq_len: ${max_seq_len}
    allow_pad_trimming: false
    decoder_only_format: true
    # # Use `python llmfoundry/data/packing.py --yaml-path /path/to/this/yaml/ ...`
    # # to profile this run's optimal packing_ratio as it depends on GPU count,
    # # batch size, sequence length
    packing_ratio: 13 # padding=0.36%, waste=0.79%
    shuffle: true
  drop_last: false
  num_workers: 8
  pin_memory: false
  prefetch_factor: 2
  persistent_workers: true
  timeout: 0

eval_loader:
  name: finetuning
  dataset:
    hf_name: mosaicml/dolly_hhrlhf
    split: test
    max_seq_len: ${max_seq_len}
    allow_pad_trimming: false
    decoder_only_format: true
    packing_ratio: 13
    shuffle: false
  drop_last: false
  num_workers: 8
  pin_memory: false
  prefetch_factor: 2
  persistent_workers: true
  timeout: 0

# Optimization
scheduler:
  name: linear_decay_with_warmup
  t_warmup: 20ba
  alpha_f: 0

optimizer:
  name: decoupled_adamw
  lr: 1e-4
  betas:
  - 0.9
  - 0.999
  eps: 1.0e-8
  weight_decay: 0.0

# FSDP
fsdp_config:
  sharding_strategy: FULL_SHARD
  mixed_precision: FULL
  activation_checkpointing: true
  activation_checkpointing_reentrant: false
  activation_cpu_offload: false
  limit_all_gathers: true
  verbose: false

# Logging
progress_bar: false
log_to_console: true
console_log_interval: 1ba

loggers:
  tensorboard: {"log_dir": "my_logs"}

# Checkpoint to local filesystem or remote object store
save_interval: 1ep
save_num_checkpoints_to_keep: 1  # Important, this cleans up checkpoints saved to DISK
save_folder: output_dir/{run_name}/checkpoints
save_overwrite: true
```

This was tested using the following command:

```
sparsify.run sparse-transfer --use-case finetune --data ../src/sparsify/auto/samples/finetune_llmfoundry_sample.yaml
```

### CODE REVIEW

File: src/sparsify/auto/tasks/finetune/finetune.py [hunk 1]
```diff
@@ -207,6 +239,14 @@ def _build_trainer(self) -> Trainer:
 
         tokenizer = build_tokenizer(self._train_config.tokenizer)
         model = self._build_model(tokenizer)
+        algorithms = []
+
+        # If a load_path is provided, try loading weights from the provided path
+        if self._train_config.get("load_path"):
+            model, algorithm = self._load_weights_and_attach_masks(model)
```
Comment (bfineran) [inline src/sparsify/auto/tasks/finetune/finetune.py:246]: where were weights previously loaded? are we loading twice now or is this because weights don't get loaded until later with composer?

File: src/sparsify/auto/tasks/finetune/helpers.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    already pruned
+    """
+    for _, module in model.named_children():
+        if isinstance(module, torch.nn.Linear):
+            mask = torch.where(
+                module.weight == 0,
+                torch.tensor(0, dtype=torch.uint8),
+                torch.tensor(1, dtype=torch.uint8),
+            )
+            module.register_buffer("mask", mask, persistent=False)
```
Comment (bfineran) [inline src/sparsify/auto/tasks/finetune/helpers.py:58]: let's make the name a little more specific to avoid conflicts with a common name, maybe `constant_pruning_mask`

File: src/sparsify/auto/tasks/finetune/helpers.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    def match(self, event, state):
+        return event == Event.BATCH_END
+
+    @torch.no_grad()
+    def apply(self, event, state, logger):
+        def mask_weights(module):
+            if hasattr(module, "mask"):
+                module.weight *= module.mask
+
+        state.model.apply(mask_weights)
```
Comment (bfineran) [inline src/sparsify/auto/tasks/finetune/helpers.py:40]: nice

**Review by bfineran — COMMENTED** [review]
looks clean, see comment

File: src/sparsify/auto/tasks/finetune/finetune.py (hunk 1 above)
Comment (dsikka) [reply src/sparsify/auto/tasks/finetune/finetune.py]: Previously, calling the `_build_model` function would've loaded the pretrained weights. Now with this new loading function, we check if a local path is provided. If that's the case, we don't load the weights from hugging face, we just provide the skeleton for the model that is needed to load the state dict. If the user does not provide the local path however, then we pull in the hugging face weights. This is basically toggled with the `pretrained` flag in the config file.

**Review by bfineran — APPROVED** [review]

**Comment by dsikka** [comment]
Last few commits update to make sure the installation works within sparsify and the rest of its requirements

llm-foundry has a requirement for accelerate which is incompatible with our transformers. Therefore, this moves sparseml[transformers] to the nm tag instead of making it a required installation. This allows us to install the correct accelerate for llm-foundry with the llm tag and the correct accelerate for sparseml[transformers] with the nm tag. Also, because our requirements are broad, there is an issue with older nightly versions. setup.py has been updated to look at more recent versions.

File: setup.py [hunk 4]
```diff
@@ -70,11 +73,11 @@ def _setup_package_dir() -> Dict:
 
 
 def _setup_install_requires() -> List:
-    return _nm_deps + _deps
+    return _deps
 
 
 def _setup_extras() -> Dict:
-    return {"dev": _dev_deps, "llm": _llm_deps}
+    return {"dev": _dev_deps, "_nm_deps": _nm_deps, "llm": _llm_deps}
```
Comment (bfineran) [inline setup.py:80]: do we need to update any core flows to install `_nm_deps`? `sparseml` is used by almost every pathway, no?

Comment (dsikka) [reply setup.py:80]: yup that's why I just moved `sparseml[transformers]` to `_nm_deps` as the problem is with `accelerate`. I can update the name `_nm_deps` to be more reflective. Added a try/except in the init for the `transformers` task imports

Comment (bfineran) [reply setup.py:80]: this is fine after speaking with product

**Review by bfineran — APPROVED** [review]
